src/exp/list_exp.rs

```diff
@@ -0,0 +1,752 @@
+            ExpressionArgument::Context(ctx.to_vec()),
+        ];
+        ListExpression::add_write(bin, ctx, args)
+    }
+
+    /// Create expression that sorts list according to sortFlags.
+    pub fn sort(
+        sort_flags: ListSortFlags,
+        bin: FilterExpression,
+        ctx: &[CdtContext],
+    ) -> FilterExpression {
+        let args = vec![
+            ExpressionArgument::Value(Value::from(ListExpOp::SORT as i64)),
+            ExpressionArgument::Value(Value::from(sort_flags as u8)),
+            ExpressionArgument::Context(ctx.to_vec()),
+        ];
+        ListExpression::add_write(bin, ctx, args)
+    }
+
+    /// Create expression that removes list items identified by value.
+    pub fn remove_by_value(
+        value: FilterExpression,
+        bin: FilterExpression,
+        ctx: &[CdtContext],
+    ) -> FilterExpression {
+        let args = vec![
+            ExpressionArgument::Value(Value::from(ListExpOp::RemoveByValue as i64)),
+            ExpressionArgument::Value(Value::from(ListReturnType::None as u8)),
+            ExpressionArgument::FilterExpression(value),
+            ExpressionArgument::Context(ctx.to_vec()),
+        ];
+        ListExpression::add_write(bin, ctx, args)
+    }
+
+    /// Create expression that removes list items identified by values.
+    pub fn remove_by_value_list(
+        values: FilterExpression,
+        bin: FilterExpression,
+        ctx: &[CdtContext],
+    ) -> FilterExpression {
+        let args = vec![
+            ExpressionArgument::Value(Value::from(ListExpOp::RemoveByValueList as i64)),
+            ExpressionArgument::Value(Value::from(ListReturnType::None as u8)),
+            ExpressionArgument::FilterExpression(values),
+            ExpressionArgument::Context(ctx.to_vec()),
+        ];
+        ListExpression::add_write(bin, ctx, args)
+    }
+
+    /// Create expression that removes list items identified by value range (valueBegin inclusive, valueEnd exclusive).
+    /// If valueBegin is null, the range is less than valueEnd. If valueEnd is null, the range is
+    /// greater than equal to valueBegin.
+    pub fn remove_by_value_range(
+        value_begin: Option<FilterExpression>,
+        value_end: Option<FilterExpression>,
+        bin: FilterExpression,
+        ctx: &[CdtContext],
+    ) -> FilterExpression {
+        let mut args = vec![
+            ExpressionArgument::Context(ctx.to_vec()),
+            ExpressionArgument::Value(Value::from(ListExpOp::RemoveByValueInterval as i64)),
+            ExpressionArgument::Value(Value::from(ListReturnType::None as u8)),
+        ];
+        if let Some(val_beg) = value_begin {
+            args.push(ExpressionArgument::FilterExpression(val_beg));
+        } else {
+            args.push(ExpressionArgument::FilterExpression(Expression::nil()));
+        }
+        if let Some(val_end) = value_end {
+            args.push(ExpressionArgument::FilterExpression(val_end));
+        }
```

Comment:
What would happen if someone passes both `value_begin` and `value_end` as `None`? Same for the other `*_by_value_range` functions below.

Reply:
I cant test it right now because i cant reach my test Server but i guess that would lead to Server ParameterError. The encoder would not fail.

Reply:
I just tested it. That results in a statement like "SELECT * FROM set".
`
        Expression::eq(
            ListExpression::get_by_value_range(
                ListReturnType::Count,
                None,
                None,
                Expression::list_bin("bin".to_string()),
                &[]),
            Expression::int_val(4)
        )
`
will return all key/value pairs in the set where all items have a list with 4 values.